crates/oxc_ast/src/ast/js.rs

```diff
@@ -1335,6 +1342,7 @@ impl<'a> FormalParameters<'a> {
 #[derive(Debug, PartialEq, Hash)]
 pub struct FunctionBody<'a> {
     pub span: Span,
+    pub is_single_expression: bool,
     pub directives: Vec<'a, Directive<'a>>,
     pub statements: Vec<'a, Statement<'a>>,
 }
```

Comment:
Added to distinguish between
```
() => x // returns x
```
and
```
() => { x } // does not return
```

Reply:
Can this be derived and implemented as a method? We don't add this kind of "state" to ASTs.

Reply:
Should we make a separate AST type for `() => x` case? Currently there's no way to derive the two cases

Reply:
https://github.com/Boshen/oxc/blob/4a447bca4f7bd1af2f3df654b06cf8a010fc17b2/crates/oxc_ast/src/ast/js.rs#L1247

`func.is_expression() && func.expression`

Reply:
On second thought, we could also de-sugar `() => x` in the parsing stage to be `() => { return x }` so there's no special case.

Reply:
> On second thought, we could also de-sugar `() => x` in the parsing stage to be `() => { return x }` so there's no special case.

I consider them to be distinct and should be preserved in the AST, otherwise our printer/formatter will encounter issues.

For a single expression, the `statements` field is irrelevant.

Therefore, should we improve the AST by designing an enum?
The `FunctionBody` should be `FunctionBlock | FunctionExpression`.

Reply:
> On second thought, we could also de-sugar `() => x` in the parsing stage to be `() => { return x }` so there's no special case.

Unfortunately no, the parser doesn't alter the syntax. If we were to implement a full compiler, then a MIR is the best place to do this.

Reply:
> https://github.com/Boshen/oxc/blob/4a447bca4f7bd1af2f3df654b06cf8a010fc17b2/crates/oxc_ast/src/ast/js.rs#L1247
> 
> `func.is_expression() && func.expression`

Yes I just found out about that 🤦‍♂️

Reply:
> Therefore, should we improve the AST by designing an enum?

This part of the AST aligns with acorn / babel so we won't change it for now.

Reply:
> > On second thought, we could also de-sugar `() => x` in the parsing stage to be `() => { return x }` so there's no special case.
> 
> I consider them to be distinct and should be preserved in the AST, otherwise our printer/formatter will encounter issues.
> 
> For a single expression, the `statements` field is irrelevant.
> 
> Therefore, should we improve the AST by designing an enum? The `FunctionBody` should be `FunctionBlock | FunctionExpression`.

I checked the current Abstract Syntax Tree design, although it may generate structures that do not exist in JavaScript, its overall concise design is positively pleasing to me. Therefore, kindly dismiss my proposal.